snowchange.py

```diff
@@ -222,6 +227,32 @@ def apply_change_script(script, change_history_table, autocommit, verbose):
   query = "INSERT INTO {0}.{1} (VERSION, DESCRIPTION, SCRIPT, SCRIPT_TYPE, CHECKSUM, EXECUTION_TIME, STATUS, INSTALLED_BY, INSTALLED_ON) values ('{2}','{3}','{4}','{5}','{6}',{7},'{8}','{9}',CURRENT_TIMESTAMP);".format(change_history_table['schema_name'], change_history_table['table_name'], script['script_version'], script['script_description'], script['script_name'], script['script_type'], checksum, execution_time, status, os.environ["SNOWFLAKE_USER"])
   execute_snowflake_query(change_history_table['database_name'], query, autocommit, verbose)
 
+# snowchange uses Jinja style variable references of the form "{{ variablename }}"
+# See https://jinja.palletsprojects.com/en/2.11.x/templates/
+# Variable names follow Python variable naming conventions
+def count_variable_references(content):
+  result = re.findall(r'\{\{ ([A-Za-z_]{1}[A-Za-z0-9_]*) \}\}', content)
+
+  if result is None:
+    return 0
+  else:
+    return len(result)
+
+def replace_variables_references(content, vars, verbose):
+  # First check to see if there are any variables to replace
+  if count_variable_references(content) == 0:
+    return content
+
+  # Loop through each variable and replace it
+  for key in vars:
+    content = re.sub(r"\{\{ " + key + r" \}\}", vars[key], content)
+
+  # If there are any leftover variables through an error now
+  if count_variable_references(content) > 0:
+    raise ValueError("Found variables remaining in script content")
+
+  return content
+
```

Comment:
I think it'd probably be better to use a [safe_substitute ](https://docs.python.org/3/library/string.html#string.Template.safe_substitute) approach instead of a manual regexp solution.

There's a couple of examples [here](https://stackoverflow.com/questions/34360603/python-template-safe-substitution-with-the-custom-double-braces-format) where string.Template has the delimiter customised.

Reply:
Alright, updated the substitution approach to use the string Template approach. Although I used substitute instead because I want the script to fail if there are left over variables in the script (since they'll for sure fail in Snowflake).

Also, I removed the region parameter as it's been deprecated in the Snowflake Python driver (https://docs.snowflake.com/en/user-guide/python-connector-api.html#connect).

Reply:
Nice! looks great